core/src/main/scala/org/apache/spark/sql/delta/commands/merge/MergeIntoMaterializeSource.scala

```diff
@@ -0,0 +1,439 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.commands.merge
+
+import java.util.UUID
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.sql.delta.{DeltaLog, DeltaTable}
+import org.apache.spark.sql.delta.files.TahoeFileIndex
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+import org.apache.spark.SparkException
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{AttributeSet, Expression, ExpressionSet, Literal}
+import org.apache.spark.sql.catalyst.optimizer.{EliminateResolvedHint, JoinSelectionHelper}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.execution.{LogicalRDD, SQLExecution}
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.execution.metric.SQLMetric
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Trait with logic and utilities used for materializing a snapshot of MERGE source
+ * in case we can't guarantee deterministic repeated reads from it.
+ *
+ * We materialize source if it is not safe to assume that it's deterministic
+ * (override with MERGE_SOURCE_MATERIALIZATION).
+ * Otherwise, if source changes between the phases of the MERGE, it can produce wrong results.
+ * We use local checkpointing for the materialization, which saves the source as a
+ * materialized RDD[InternalRow] on the executor local disks.
+ *
+ * 1st concern is that if an executor is lost, this data can be lost.
+ * However, when Spark executor decomissioning API is used, it should attempt to move this
+ * materialized data safely out before removing the executor.
+ *
+ * 2nd concern is that if an executor is lost for another reason (e.g. spot kill), we will
+ * still lose that data. To mitigate that, we implement a retry loop.
+ * The whole Merge operation needs to be restarted from the beginning in this case.
+ * When we retry, we increase the replication level of the materialized data from 1 to 2.
+ * (override with MERGE_SOURCE_MATERIALIZATION_RDD_STORAGE_LEVEL_RETRY).
+ * If it still fails after the maximum number of attempts (MERGE_MATERIALIZE_SOURCE_MAX_ATTEMPTS),
+ * we record the failure for tracking purposes.
+ *
+ * 3rd concern is that executors run out of disk space with the extra materialization.
+ * We record such failures for tracking purpuses.
+ */
+trait MergeIntoMaterializeSource extends DeltaLogging {
+  import MergeIntoMaterializeSource._
+
+  /**
+   * Prepared Dataframe with source data.
+   * If needed, it is materialized, @see prepareSourceDF
+   */
+  protected var sourcePlanDF: Dataset[Row] = null
```

Comment:
nit: i dont like having `plan` and `DF` in the same name. DF is already a plan. rather `plan` using means the object actually a LogicalPlan or SparkPlan. so I suggest `sourceDF` or `preparedSourceDF`

Reply:
also, why is this using `null` and the `materializedSourceRDD` below using Option[..] = None?
best to be consistent